gh-760: adding colour display for theme properties

CristianGimenez opened the pull request:
On property display, added a disabled input of type color to show the selected colour of the property value.
On property edit/add if the key of the property contains "theme.color" show 2 inputs, the user can use both of them, one is an input text where the hex code of the colour can be pasted, the other is an input of type colour so we can have an html colour picker.

As discussed in a conversation with @NigelPalmer,  in the issue description we say to check the key of the property and look for those which starts with "explorer.theme.color" but after some thought we changed that to be properties that contain "theme.color" since in the future this might be used to also theme mdm-ui.

Closes #760

CristianGimenez committed:
gh-760: adding colour display for theme properties

CristianGimenez committed:
gh-760: changing startsWith for includes

pjmonks reviewed (changes requested):
Having tested this it seems to work well. And although a colour picker control was originally out of scope, the fact that one is present in the add/edit property screen is very useful.

There is one big issue linked to the Mauro Data Explorer theme though, in that the current API keys used for the MDE theme do not match the expression "theme.color", so all current keys listed for the MDE do not show these colour options. Issue #760 wasn't updated with the correct task description I think - look at MauroDataMapper-Plugins/mdm-plugin-explorer#18 for the correct table of possible values.

This means some of this logic will have to change accordingly to match.

CristianGimenez committed:
gh-760:  creating pure pipe for checking regex

CristianGimenez commented:
#Thanks @pjmonks, you were right, the issue wasn't up to date. After reviewing the list of properties added in [MauroDataMapper-Plugins/mdm-plugin-explorer#18](https://github.com/MauroDataMapper-Plugins/mdm-plugin-explorer/issues/18)... the pattern I've seen is that all of the colour properties there, have a part of the key that includes "theme" and later another part that includes "color", therefore, I've changed the using of `String.includes` to a regular expression match. Any key that contains:

- any or no characters (not numeric), followed by
- the string "theme" followed by,
- any or no characters (not numeric), followed by
- the string "color", followed by,
- any or no characters (not numeric)

will pass the regex expression which is` \D*theme\D*color\D*` by the way.

As discussed, I didn't see how to avoid having a call to a function in the template, which causes the function to be called each time Angular checks for changes (which can be a lot). Reading around I've found [this article ](https://medium.com/showpad-engineering/why-you-should-never-use-function-calls-in-angular-template-expressions-e1a50f9c0496#:~:text=The%20solution%3A%20how%20to%20avoid%20unnecessary%20function%20calls&text=A%20first%20strategy%20to%20reduce,pipe's%20input%20does%20not%20change.) that states pure pipes as an alternative. For the tests I've made it seems not to be called in a regular basis. So I have:
- created a pipe
- added tests for that pipe to check that the regular expression works as expected
- added the pipe to the admin module
- substituted the `String.includes` calls with the newly created pipe

I think this is ready to review again. Thanks for your help!

pjmonks reviewed (dismissed):
This looks better now. As far as I can tell, the table is not being constantly rendered thanks to the pipe used, which gets around having logic be part of the data binding 👍 

![image](https://user-images.githubusercontent.com/3219480/221915764-928e7ce5-5fe5-48b5-bb60-f93f04aed128.png)

Approved ✔️

jamesrwelch commented:
Could I make one minor styling request?  Could the colour square be horizontally aligned on the right of the text?  Or on the left of the text as I suspect that might be easier?  Where you have short colours (`#fff`), and long colours, (`#000000`), the squares wiggle all over the place :)

CristianGimenez committed:
gh-760: style changes

CristianGimenez commented:
Updated after the comments from @jamesrwelch, now the span with the colour text or the input text for the edit page, have a fixed width of 100px, that will make the coloured square to be in place even with smaller text lengths

CristianGimenez committed:
gh-760: bug fix

pjmonks reviewed (dismissed):
Colour values in table now look aligned:

![image](https://user-images.githubusercontent.com/3219480/222450608-05550776-0a55-4def-943c-821b4144f611.png)

Approved ✔️

pjmonks committed:
gh-760 eslint fixes

jamesrwelch commented:
Ooh, very aligned, I like it 👍 
Not really a discussion for this PR, but perhaps a wider one - `px` vs. `em`/`rem`?

pjmonks commented:
@jamesrwelch Could you merge this please? Everything is approved but it's waiting for the license/cla check for some reason.
